osu.Framework/Graphics/Textures/TextureLockerBitmap.cs

```
@@ -0,0 +1,29 @@
+﻿// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu-framework/master/LICENCE
+
+using System;
+using System.Drawing;
+using System.Drawing.Imaging;
+
+namespace osu.Framework.Graphics.Textures
+{
+    public class TextureLockerBitmap : ITextureLocker
+    {
+        private readonly Bitmap bitmap;
+
+        private readonly BitmapData data;
+
+        public IntPtr DataPointer => data.Scan0;
+
+        public TextureLockerBitmap(Bitmap bitmap, Rectangle region)
+        {
+            this.bitmap = bitmap;
+            data = bitmap.LockBits(region, ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);
+        }
+
+        public void Dispose()

```

Any reason why `Dispose` isn't called within a finalizer? Does `LockBits` somehow prohibit garbage collection? (Same argument for TextureLockerByteArray)

---

hmm, we can add a finalizer but it should probably throw an exception or something. these classes are not ever intended to be GC finalized as they handle underlying unmanaged memory references.

in this case many `TextureLockerBitmaps`s can refer to the same bitmap, but locking bits can only happen once at a time (a second call without first unlocking will break). if it happened that this was left to the finaliser we would already be in a bad state.

---

`bool isDisposed` could help ensure single-disposal. Unfortunately, throwing within a finalizer isn't permitted, but at least we can prevent memory leaks by having this fail-safe.

That said, doesn't `GC.SuppressFinalize(this);` already prevent the finalizer from running in case `Dispose` has been manually called beforehand?

---

Ah I didn't mean it that way.

Just these classes are only used in very controlled places. Finalizers will never be used and if they are used the class is being used incorrectly already.

---

We can make all these classes internal if that helps.

---

Putting a Debug.Assert call within the finalizer may be acceptable then. Not sure if assertions are in any way more acceptable than throwing exceptions, but probably worth a shot? `Debug.Fail` may be a more explicit option over `Debug.Assert(false)`.